feat(api): implement websocket endpoint for streaming inference

osamasydysf opened the pull request:
This PR introduces a new FastAPI-based WebSocket API to support real-time streaming inference for the Quran Muaalem application.

- **WebSocket Server**: Implements a complete WebSocket server with a custom streaming protocol and
  session management. A new CLI entrypoint is provided for running the server.
- **Rich Inference Data**: The server now includes a detailed `phonetizer_out` field in both
  intermediate and final responses. The serialization utility was enhanced to support complex data
  types like dataclasses, tensors, and numpy types.
- **Data Integrity**: A bug was fixed to ensure that any partial audio data remaining in the
  buffer is flushed when a stream ends, preventing data loss.
- **Testing**: Comprehensive end-to-end and buffer reset tests have been added using pytest to
  validate the new WebSocket functionality.

osamasydysf committed:
Add FastAPI WebSocket API server and tests  Introduces a FastAPI-based WebSocket API server for Quran Muaalem, including a streaming inference protocol and session management. Adds API dependencies and a new CLI entrypoint in pyproject.toml. Provides end-to-end and buffer reset WebSocket tests with audio streaming using pytest.

Add FastAPI WebSocket API server and tests

Introduces a FastAPI-based WebSocket API server for Quran Muaalem, including a streaming inference protocol and session management. Adds API dependencies and a new CLI entrypoint in pyproject.toml. Provides end-to-end and buffer reset WebSocket tests with audio streaming using pytest.

osamasydysf committed:
fix(websocket): flush partial audio from buffer on stream end

- update the websocket 'end' message handler to flush remaining audio from the buffer
- this ensures that partial audio segments smaller than the processing window are not lost
- modify RollingBuffer helpers to better handle partial audio data
- guarantees all audio is processed and sent to the client before the connection closes

osamasydysf committed:
feat(websocket): include phonetizer_out in inference websocket responses

- add a serialized 'phonetizer_out' field to both final and intermediate inference messages
- enhance the `_to_serializable` utility to better handle dataclasses, pydantic models, tensors, numpy types, and enums
- update tests to verify the presence and structure of the new 'phonetizer_out' field

osamasydysf committed:
docs(api): update inference response JSON documentation

- add documentation for the 'final' and 'phonetizer_out' fields in the inference response
- clarify the purpose and structure of these fields in the websocket message format

hossam-96 commented on src/fastapi_server.py:
```diff
@@ -0,0 +1,323 @@
+async def _startup() -> None:
+    # Load the model once
+    device = "cuda" if torch.cuda.is_available() else "cpu"
+    app.state.muaalem = Muaalem(device=device)
```
device priorities:
1. cuda
2. mps (for Apple silicone)
3. cpu

dtype priorities:
1. bf16 (if supported)
2. float32

hossam-96 commented on src/fastapi_server.py:
```diff
@@ -0,0 +1,323 @@
+            await ws.close(code=1002)
+            return
+        session.configure(cfg)
+        await ws.send_text(json.dumps({"type": "ready", "sampling_rate": session.sr, "audio_format": session.audio_format}))
+
+        while True:
```
Why keeping the worker idle instead of returning and wait for an audio event ?

osamasydysf replied:
We're not actually keeping a thread “busy.” That loop is event-driven.

- await ws.receive() suspends the coroutine and yields control back to the event loop. While no audio arrives, it consumes essentially no CPU and doesn’t block other connections.
- If you “return,” the handler exits and the WebSocket closes; there’s no way to “wait for an audio event” without an active connection.

hossam-96 commented on src/fastapi_server.py:
```diff
@@ -0,0 +1,323 @@
+
+                out = outs[0]
+                result: Dict[str, Any] = {
+                    "phonemes": {
+                        "text": getattr(out.phonemes, "text", None),
+                        "probs": _to_serializable(getattr(out.phonemes, "probs", None)),
+                        "ids": _to_serializable(getattr(out.phonemes, "ids", None)),
+                    },
+                    "sifat": [_to_serializable(s) for s in getattr(out, "sifat", [])],
```
This is the raw output of the model, the app expects an output similar to the shown in terminal using the `explain` function, check with Mohamed on this

osamasydysf replied:
This is returning the predicted phonemes and sifat in `result` and the true phonemes and sifat in `phonetizer_out` these are the inputs for the `explain` function so that mohammed can create his own `explain` function in flutter side.

hossam-96 commented on src/fastapi_server.py:
```diff
@@ -0,0 +1,323 @@
+        self.aya_ref_text = uthmani_ref
+        self.phonetizer_out = quran_phonetizer(uthmani_ref, moshaf, remove_spaces=True)
+
+    def decode_binary_audio(self, data: bytes) -> np.ndarray:
+        # Expect mono PCM. For pcm16le: int16 little-endian, for float32: IEEE 754 little-endian
+        if self.audio_format == "float32":
+            arr = np.frombuffer(data, dtype='<f4')  # little-endian float32
+            return arr.astype(np.float32)
+        # default pcm16
+        arr = np.frombuffer(data, dtype='<i2')
+        return (arr.astype(np.float32) / 32768.0)
+
+    def decode_json_audio(self, payload: Dict[str, Any]) -> np.ndarray:
+        b64 = payload.get("chunk_base64")
+        if not b64:
+            return np.array([], dtype=np.float32)
+        raw = base64.b64decode(b64)
+        return self.decode_binary_audio(raw)
```
Check if `librosa` package is faster, you don't need to handle two types of input, just choose whichever is faster and send it from the app.

osamasydysf replied:
- Librosa isn’t relevant here. It’s for file I/O/resampling, not decoding raw PCM frames. It will be slower and add overhead.
- For streaming bytes, numpy.frombuffer is already the fastest/lowest-overhead path.
- Prefer pcm16le: halves bandwidth vs float32 and decode is a single cast/scale.
- float32 offers no meaningful speedup; it doubles network payload.
- I've dropped support for float32 and json/base64.

osamasydysf committed:
perf(model): add device and dtype selection for model loading

- introduce helper functions to select the optimal device (cuda, mps, or cpu)
- automatically select bfloat16 for dtype if supported, otherwise fall back to float32
- update fastapi startup event to use these selections when loading the muaalem model
- this improves compatibility and performance across different hardware environments

osamasydysf committed:
refactor(api)!: simplify audio input to binary pcm16le only

- remove support for base64-encoded and float32 audio input formats
- the server now exclusively accepts binary pcm16le audio frames
- this change simplifies the protocol and reduces code complexity
- update documentation and remove unused code paths to reflect the change

BREAKING CHANGE: The server no longer accepts base64 or float32 audio. Clients must send audio as binary PCM16LE frames.
